xain_fl/grpc/participant.py

```diff
@@ -0,0 +1,377 @@
+            )
+            # which one was it?
+            return self.state
+
+
+def transit(st, beat_reply):
+    """Participant state transition function on a heartbeat response. Updates the
+    state record `st`.
+
+    Args:
+        st (obj:`StateRecord`): Participant state record to update.
+        beat_reply (obj:`coordinator_pb2.HeartbeatReply`): Heartbeat from Coordinator.
+
+    """
+    msg, r = beat_reply.state, beat_reply.round
+    with st.cv:
+        if st.state == ParState.WAITING_FOR_SELECTION:
+            if msg == coordinator_pb2.State.Value("ROUND"):
+                st.state = ParState.TRAINING
+                st.round = r
+                st.cv.notify()
+            elif msg == coordinator_pb2.State.Value("FINISHED"):
+                st.state = ParState.DONE
+                st.cv.notify()
+        elif st.state == ParState.POST_TRAINING:
+            if msg == coordinator_pb2.State.Value("STANDBY"):
+                # not selected
+                st.state = ParState.WAITING_FOR_SELECTION
+                # prob ok to keep st.round as it is
+                st.cv.notify()
+            elif msg == coordinator_pb2.State.Value("ROUND") and r == st.round + 1:
+                st.state = ParState.TRAINING
+                st.round = r
+                st.cv.notify()
+            elif msg == coordinator_pb2.State.Value("FINISHED"):
+                st.state = ParState.DONE
+                st.cv.notify()
+
+
+def message_loop(chan, st, terminate):
+    """Periodically sends (and handles) heartbeat messages in a loop.
+
+    Args:
+        chan: gRPC channel to Coordinator.
+        st (obj:`StateRecord`): Participant state record.
+        terminate (obj:`threading.Event`): Event to terminate message loop.
+    """
+    coord = coordinator_pb2_grpc.CoordinatorStub(chan)
+    while not terminate.is_set():
+        start = datetime.today()
+        print(f"sending heartbeat {start}")
+        if st.state in [ParState.WAITING_FOR_SELECTION, ParState.POST_TRAINING, ParState.DONE]:
+            state = coordinator_pb2.State.Value("READY")
+        else:
+            state = coordinator_pb2.State.Value("TRAINING")
+        req = coordinator_pb2.HeartbeatRequest(state=state, round=st.round)
+        reply = coord.Heartbeat(req)
+        transit(st, reply)
+        end = datetime.today() - start
+        print(f"ended sending heartbeat {end}")
+        time.sleep(HEARTBEAT_TIME)
+
+
+def go(part: Participant, coordinator_address: str):
+    """Top-level function for the Participant state machine.
+
+    After rendezvous and heartbeat initiation, the Participant is
+    WAITING_FOR_SELECTION. When selected, it moves to TRAINING followed by
+    POST_TRAINING. If selected again for the next round, it moves back to
+    TRAINING, otherwise it is back to WAITING_FOR_SELECTION.
+
+    Args:
+        part (obj:`Participant`): Participant object for training computation.
+    """
+
+    # channel options
+    options = [
```

Comment:
These options remove the size limit of sent and received messages. Right now the reason is for the messages that send numpy arrays. In the future the numpy arrays may not be shared through gRPC or if they are we need to choose and bound the maximum message size.